### PR TITLE
Support for remote Chromium/Chrome instance

### PR DESCRIPTION
**Pull request by videowala**
Added a new config parameter(--remote-url). If the remote URL is given *puppeteer.connect()* will be used intead of *puppeteer.launch()*

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -119,11 +119,23 @@ const makeFileDirectoryIfNeeded = function (filepath) {
   }
 };
 
+
+const getBorwser = async function(url, launchOptions) {
```
Comment (tungs) [inline lib/utils.js]: In an effort to still support Node 6.4, I'd avoid using `async` functions. It turns out to be more elegant anyway.

```suggestion
const getBrowser = function(config, launchOptions) {
```

File: lib/utils.js [hunk 2]
```diff
@@ -119,11 +119,23 @@ const makeFileDirectoryIfNeeded = function (filepath) {
...
+  if(url){
```
Comment (tungs) [inline lib/utils.js]:
```suggestion
  if (config.remoteUrl) {
```

File: lib/utils.js [hunk 3]
```diff
@@ -119,11 +119,23 @@ const makeFileDirectoryIfNeeded = function (filepath) {
...
+    var queryString = Object.keys(launchOptions).map(key => key + '=' + launchOptions[key]).join('&');
```
Comment (tungs) [inline lib/utils.js]: Can you point to where query strings are described by the browser endpoint? You may want to use encodeURIComponent on the key or the value.

File: lib/utils.js [hunk 4]
```diff
@@ -119,11 +119,23 @@ const makeFileDirectoryIfNeeded = function (filepath) {
...
+    let remote = url+"?"+queryString
```
Comment (tungs) [inline lib/utils.js]:
```suggestion
    var remote = config.remoteUrl + '?' + queryString;
```

File: lib/utils.js [hunk 5]
```diff
@@ -119,11 +119,23 @@ const makeFileDirectoryIfNeeded = function (filepath) {
...
+    return await puppeteer.connect({ browserWSEndpoint: remote })
```
Comment (tungs) [inline lib/utils.js]:
```suggestion
    return puppeteer.connect({ browserWSEndpoint: remote });
```

File: lib/utils.js [hunk 6]
```diff
@@ -119,11 +119,23 @@ const makeFileDirectoryIfNeeded = function (filepath) {
...
+  }else{
```
Comment (tungs) [inline lib/utils.js]:
```suggestion
  } else {
```

File: lib/utils.js [hunk 7]
```diff
@@ -119,11 +119,23 @@ const makeFileDirectoryIfNeeded = function (filepath) {
...
+    return await puppeteer.launch(launchOptions)
```
Comment (tungs) [inline lib/utils.js]:
```suggestion
    return puppeteer.launch(launchOptions);
```

File: index.js [hunk 8]
```diff
@@ -104,7 +104,8 @@ module.exports = function (config) {
     args: config.launchArguments || []
   };
 
-  return puppeteer.launch(launchOptions).then(function (browser) {
+
+  return getBorwser(config.remoteUrl,launchOptions).then(function (browser) {
```
[the remark this answers is not in the document]
Comment (videowala) [reply index.js]: Yeah, this makes it more flexible.

File: lib/utils.js (hunk 3 above)
Comment (videowala) [reply lib/utils.js]: It is specific to Browserless. I am not sure if the generic to puppeteer.
https://docs.browserless.io/docs/chrome-flags.html

**Comment by videowala** [comment]
@tungs Thanks for reviewing the code and suggestions. I have applied all the suggestion.

**Pull request merged**